### PR TITLE
HackStudio: Add JS expression evaluation to the debugger

### PR DESCRIPTION
**Pull request by FalseHonesty**
This implements a dialog that can be used to evaluate a JS expression in the HackStudio's Debugger context. It also implements simple C++ Variable <-> JS Value conversion, allowing for JS expressions to read/write variables in the debugger scope.

To access the expression evaluator, start debugging a process, wait for a breakpoint to hit, and then right click the editor and select the evaluate expression context menu option. (Note: I took the idea for how to reach the expression evaluator from the JetBrains IDEs). From there, you can simply reference any variable in the debugger variables table by name and execute arbitrary expressions with their values!

Thanks to @mattco98 for his help in making some of the LibJS interoperability work well :)

preview

### CODE REVIEW

File: Userland/Libraries/LibJS/Runtime/VM.cpp [hunk 1]
```diff
@@ -193,8 +193,11 @@ Value VM::get_variable(const FlyString& name, GlobalObject& global_object)
 
         for (auto* scope = current_scope(); scope; scope = scope->parent()) {
             auto possible_match = scope->get_from_scope(name);
-            if (possible_match.has_value())
+            if (possible_match.has_value()) {
                 return possible_match.value().value;
+            } else if (exception()) {
+                return {};
+            }
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/VM.cpp:200]: I'd prefer the usual "exception check first, return immediately if any" style:

```suggestion
            if (exception())
                return {};
            if (possible_match.has_value())
                return possible_match.value().value;
```

Comment (FalseHonesty) [reply Userland/Libraries/LibJS/Runtime/VM.cpp]: Ah gotcha, thanks for the tip!

**Comment by FalseHonesty** [comment]
Since I forgot before, cc @itamar8910 :)

File: Userland/DevTools/HackStudio/Debugger/DebuggerObject.h [hunk 2]
```diff
@@ -0,0 +1,54 @@
+/*
```
Comment (itamar8910) [inline Userland/DevTools/HackStudio/Debugger/DebuggerObject.h:1]: Small nitpick: Can we call this something that indicates it has to do with JS integration and not with the debugging mechanism itself, perhaps `DebuggerGlobalJSObject`?

File: Userland/DevTools/HackStudio/Debugger/DebuggerVariableObject.h [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+#pragma once
+
+#include "DebuggerObject.h"
+#include <LibDebug/DebugInfo.h>
+#include <LibJS/Runtime/Object.h>
+
+namespace HackStudio {
+
+class DebuggerVariableObject final : public JS::Object {
```
Comment (itamar8910) [inline Userland/DevTools/HackStudio/Debugger/DebuggerVariableObject.h:36]: same thing as the `DebuggerObject`, maybe rename to `DebuggerVaraibleJSObject`?

**Review by itamar8910 — APPROVED** [review]
Sorry for the delay, I only now got to reviewing this.
Anyways, This is an awesome addition to the debugger, great stuff! :rocket:

File: Userland/DevTools/HackStudio/Debugger/DebuggerObject.h (hunk 2 above)
Comment (FalseHonesty) [reply Userland/DevTools/HackStudio/Debugger/DebuggerObject.h:1]: Yeah absolutely, good point.

**Review by itamar8910 — APPROVED** [review]

File: Userland/DevTools/HackStudio/Debugger/DebuggerGlobalJSObject.cpp [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
```
Comment (linusg) [inline Userland/DevTools/HackStudio/Debugger/DebuggerGlobalJSObject.cpp:25]: As of #6547 copyright headers don't include the full license text anymore, please update :)

File: Userland/DevTools/HackStudio/Debugger/DebuggerGlobalJSObject.h [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
```
Comment (linusg) [inline Userland/DevTools/HackStudio/Debugger/DebuggerGlobalJSObject.h:25]: same here

File: Userland/DevTools/HackStudio/Debugger/DebuggerVariableJSObject.cpp [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
```
Comment (linusg) [inline Userland/DevTools/HackStudio/Debugger/DebuggerVariableJSObject.cpp:26]: same here

File: Userland/DevTools/HackStudio/Debugger/DebuggerVariableJSObject.h [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
```
Comment (linusg) [inline Userland/DevTools/HackStudio/Debugger/DebuggerVariableJSObject.h:26]: same here

File: Userland/DevTools/HackStudio/Debugger/EvaluateExpressionDialog.cpp [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
```
Comment (linusg) [inline Userland/DevTools/HackStudio/Debugger/EvaluateExpressionDialog.cpp:25]: same here

File: Userland/DevTools/HackStudio/Debugger/EvaluateExpressionDialog.h [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
```
Comment (linusg) [inline Userland/DevTools/HackStudio/Debugger/EvaluateExpressionDialog.h:25]: same here

**Review by linusg — CHANGES_REQUESTED** [review]

**Pull request merged**